### ISSUE TITLE
TreeSAGE ?

### ISSUE THREAD

**Issue by CoteDave** [iancovert/sage#17]
Hi,

Do you think is it possible to adapt the SAGE approach using a faster approximation for tree-like models like the TreeSHAP algorithm for shapley values ?

**Comment by iancovert** [comment]
It's not something I've worked on, but I've seen at least [one paper](https://arxiv.org/abs/2109.00855) on the subject and I've thought about it before. There are a couple ways to approach it that achieve different tradeoffs between approximation quality and speed, and it would take some time to figure out what makes the most sense.

For example, at one end of the spectrum, we could leverage the tree structure when querying the model with each feature subset: the split proportions can help handle unknown feature values. This wouldn't compromise the approximation quality, and I suspect it would provide a modest speedup.

On the other end of the spectrum, you could make a very aggressive approximation and do something like multiply the per-sample TreeSHAP values by the loss gradient at the actual prediction, and then average this across samples. Intuitively, that represents whether eliminating a feature on average moves predictions in a direction that helps or hurts the loss value. This is extremely fast, but it really compromises the approximation quality. (Not so much that it's not useful - I've used it successfully for high-dimensional nonlinear feature selection, but they're hardly SAGE values at this point.)

I'm actually not an expert on TreeSHAP, so I'm not sure if there's another option in the middle that resembles a specialization of the TreeSHAP dynamic programming algorithm for SAGE values.

Is this something that would be very helpful for your use case?

**Comment by CoteDave** [comment]
Hi, thanks for the great answer!

The main use I see is that gradient boosting is very very popular and often, when working with millions of rows and undreds of features, SAGE "kernel" is very intensive and long to compute (In this case I prefer to use the global approximation of averaging absolute local Shapley values from treeSHAP for efficiency reason). I was hoping a treeSAGE approach would help to significantly reduce the computing time like it is the case for treeSHAP vs KernelSHAP.

**Comment by iancovert** [comment]
You're right, that would be pretty useful. Let me talk to someone who knows TreeSHAP very well and get back to you.

**Comment by iancovert** [comment]
Alright, I have a solution that actually requires very little effort to try! The idea is that SAGE values are the average of LossSHAP values - in case you're not familiar with that, it's when you use SHAP on a per-prediction basis to explain the loss rather than the raw prediction. (We discuss it a bit in the paper, and also in this [blog post](https://iancovert.com/blog/understanding-shap-sage/).)

Because of that, the simple approach you can do is 1) calculate per-sample LossSHAP values, 2) average those across samples. And voila, you'll have a very fast estimate of the SAGE values.

As for implementing LossSHAP, it's already supported in the shap repo - see the argument that you need to specify [here](https://github.com/slundberg/shap/blob/master/shap/explainers/_tree.py#L79) (`model_output='log_loss'`). It's only supported for `feature_dependence='independent'` (i.e., the version that's a worse approximation of the conditional distribution), so that's arguably not ideal, but it's definitely good enough to be useful.

Let me know how that sounds and if you're able to give it a try. If it's useful, I could write a wrapper around it so folks could run it from the SAGE package.

**Comment by CoteDave** [comment]
Hi Ian!

Here what I did following your advice:

shap_explainer_interv_sage = shap.TreeExplainer(model_catboost,
                                           data=X_test, 
                                           model_output='log_loss', 
                                           feature_perturbation='interventional',)

shap_val_interv_sage = shap_explainer_interv_sage.shap_values(X_test, y_test, check_additivity=False)

Seems to work! I Observed some small rank differences vs model_ouput = 'raw' soo it does give an alternative answer with the same background data and model.

However, I Also observed big rank differences using SAGE on the same dataset. A major difference is that with SAGE, I was forced to use a small sample for performance reasons. 
Based on this small experiment, I would say that model_output='log_loss' vs SAGE are actually not giving the same results (Maybe the "interventional" approach or/and the sampling are the reason why ?). 
Also, SAGE gave me some features with negatives shapley values but not the model_output='log_loss' approach.

In conclusion, when SAGE is not possible for performance reasons (big datasets), I would definitly use model_output='log_loss'as an alternative to compute global shapley values instead of the model_ouput = 'raw'.

What do you think ?

**Comment by iancovert** [comment]
What you've described makes sense! I have a couple thoughts on why the output from TreeSHAP and SAGE were different:

1. Using a different set of samples. But of course you tried addressing this, so it's not the main reason. (On a sidenote, I would have thought SAGE's time to converge wouldn't scale too significantly in the number of samples, or at least not linearly. It sounds like that's not what you observed?)
2. The way held-out features are removed shouldn't be very different. SAGE would ideally do something "observational", but the `MarginalImputer` object is actually roughly equivalent to what SHAP is doing here. (It's a little hard to tell from the code, but I think `independent` is the same as `interventional`.)
3. The main reason, I think, is that SHAP uses a cheap approximation of the Shapley values for the model's loss, whereas SAGE calculates this properly. In that sense, I would expect SAGE's output to be more reliable. However, this comes at the expense of more computation, and I suppose the tradeoff isn't always worth it.

How different were the rankings you saw? Were they at least similar for the most important features? 

And the final interpretation you gave seems right to me: that when SAGE is too costly to compute, I would probably consider using SHAP's `log_loss` option over mean absolute values with the `raw` option. Perhaps there's an easy way to test this... Have you by any chance tried something like retraining a model with the top features (e.g., the top 10), and seeing which version gives the best performance?
